tests/functional/conftest.py

```diff
@@ -0,0 +1,268 @@
+            "password": password,
+        },
+        "exporter": {
+            "port": 5000,
+            "collect_interval": int(COLLECTION_INTERVAL / 60),
+        },
+    }
+
+
+def configure_snap(version):
+    """Configure the snap."""
+    user, password, cacert, endpoint = get_juju_data()
+    config_data = generate_config_data(user, password, cacert, endpoint)
+
+    temp_config_file = os.path.join(TMP_DIR, "prometheus_juju_exporter_config.yaml")
+    with open(temp_config_file, "w") as config_file:
+        try:
+            yaml.dump(config_data, config_file)
+        except yaml.YAMLError as e:
+            logging.error(e)
+
+    snap_config_file = os.path.join(SNAP_CONFIG_DIR, version, "config.yaml")
+    assert (
+        check_call(f"sudo mv {temp_config_file} {snap_config_file}".split())
+        == 0  # noqa
+    )
+
+    assert os.path.isfile(snap_config_file)
+
+
+def get_snap_version():
+    """Get the version of the installed snap."""
+    try:
+        output = (
+            check_output(f"snap info {SNAP_NAME}".split()).decode().strip().splitlines()
+        )
+        package_info = [i for i in output if "installed:" in i]
+        version = package_info[0].split("(")[1].split(")")[0]
+    except Exception as e:
+        logging.error("Error getting snap version: %s", e)
+
+    return version
+
+
+@pytest.fixture(scope="session", autouse=True)
+def setup_snap():
+    """Install the package to the system and cleanup afterwards.
+
+    An environment variable TEST_SNAP is needed to install the snap.
+    """
+    test_snap = os.environ.get("TEST_SNAP", None)
+    if test_snap:
+        logging.info("Installing %s snap package...", test_snap)
+        assert os.path.isfile(test_snap)
+        assert (
+            check_call(f"sudo snap install --dangerous {test_snap}".split())
+            == 0  # noqa
+        )
+
+        version = get_snap_version()
+        configure_snap(version)
+
+        assert check_call(f"sudo snap start --enable {SNAP_NAME}".split()) == 0  # noqa
+    else:
+        logging.error(
+            "Could not find %s snap package for testing. Needs to build it first.",
+            SNAP_NAME,
+        )
+
+    yield test_snap
+
+    if test_snap:
+        logging.info("Removing %s snap package...", SNAP_NAME)
+        check_call(f"sudo snap stop --disable {SNAP_NAME}".split())
+        check_call(f"sudo snap remove {SNAP_NAME}".split())
+
+
+@pytest.fixture(scope="session", autouse=True)
+def setup_test_model():
+    """Create a test model for machine manipulation."""
+    test_model_name = "prometheus-juju-exporter-func-test"
+    logging.info("Creating model '%s'...", test_model_name)
+    assert check_call(f"juju add-model {test_model_name}".split()) == 0  # noqa
+
+    yield test_model_name
+
+    logging.info("Destroying model '%s'...", test_model_name)
+    check_call(
+        f"juju destroy-model --destroy-storage --force -y {test_model_name}".split()
+    )
+
+
+@pytest.fixture
```

Comment:
What's the benefit of marking this function as a fixture and then returning inner callable function? Wouldn't it be better to take this function (along with those below), put it into separate "helper" module and just use it from there?

Reply:
There's no particular benefit other than having all the "non-test" functions in one place. I agree with you that having these helper functions in a separate module would probably be better in terms of the overall structure and readability.